grai-integrations/source-metabase/src/grai_source_metabase/loader.py

```diff
@@ -0,0 +1,295 @@
+        namespace_map (Dict[int, str]): A mapping of database IDs to their corresponding namespace names.
+        question_table_map (Dict[int, int]): A mapping of question IDs to their corresponding table IDs.
+        table_db_map (Dict[int, int]): A mapping of table IDs to their corresponding database IDs.
+        question_db_map (Dict[int, int]): A mapping of question IDs to their corresponding database IDs.
+
+    """
+
+    def __init__(
+        self,
+        namespaces: Optional[Dict] = None,
+        default_namespace: Optional[str] = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+
+        self.default_namespace = default_namespace
+
+        # This line creates a list of tables by modifying each table dictionary obtained from the get_tables() method.
+        # It replaces the "schema" key with a new key "schema_name" while preserving the other key-value pairs.
+        # this is because the "schema" key is a reserved keyword in the pydantic.
+        self.tables = [
+            {**table, "schema_name": table.pop("schema")} for table in self.get_tables()
+        ]
+        self.tables_map = {table["id"]: table for table in self.tables}
+        self.dbs_map = {db["id"]: db for db in self.get_dbs()["data"]}
+        self.questions_map = {
+            question["id"]: question for question in self.get_questions()
+        }
+
+        self.namespace_map = build_namespace_map(
+            self.dbs_map, namespaces, default_namespace
+        )
+
+        self.question_table_map = {}
+        self.table_db_map = {}
+        self.question_db_map = {}
+
+    def build_lineage(self):
+        """
+        Builds the lineage information by updating the mappings and maps.
+
+        This method updates the `tables_map`, `dbs_map`, `questions_map`, `question_table_map`, `table_db_map`,
+        and `question_db_map` attributes.
+
+        """
+
+        self.question_table_map = {
+            question["id"]: question["table_id"]
+            for question in self.get_questions()
+            if question["table_id"] and question["archived"] is False
+        }
+
+        self.table_db_map = {table["id"]: table["db_id"] for table in self.tables}
+
+        for question, table in self.question_table_map.items():
+            self.question_db_map[question] = self.table_db_map[table]
+
+    def get_nodes(self) -> List[NodeTypes]:
+        """
+        Retrieves the list of nodes representing tables and questions.
+
+        Returns:
+            List[NodeTypes]: The list of nodes.
+
+        """
+
+        for question in self.questions_map.values():
+            question["namespace"] = self.namespace_map[
+                self.question_db_map[question["id"]]
+            ]
+
+        for table in self.tables_map.values():
+            table["namespace"] = self.namespace_map[self.table_db_map[table["id"]]]
```

Comment:
This errors unless the user has run `build_lineage` first. To fix this add something like an `self._lineage_ready=False` flag in the init, at the end of `build_lineage` set the `self._lineage_ready=True` then at the top of get_nodes and get_edges include

```python
if not self._lineage_ready:
  self.build_lineage()
```